feat(api): autocmd `group` can be either name or id

dhruvmanila opened the pull request:
This affects the following API functions:
- `vim.api.nvim_create_autocmd`
- `vim.api.nvim_get_autocmds`
- `vim.api.nvim_do_autocmd`

### Questions/doubts:
1. Should I create a new static function to resolve the `group` parameter value as the same logic is present in 3 different functions?
2. ~I'm not sure whether the logic for checking whether the augroup id exists is correct. If anyone can help me with it, that would be great.~

### Test plan:
- [x] check if it raises an error when augroup id does not exists
- [x] check if it raises an error if `group` is neither string or int
- [x] check if `nvim_create_autocmd` accepts group by id and name
- [x] check if `nvim_get_autocmds` accepts group by id and name
- [x] check if `nvim_do_autocmd` accepts group by id and name

closes #17552

dhruvmanila committed:
feat(api): `group` can be either string or int

This affects the following API functions:
- `vim.api.nvim_create_autocmd`
- `vim.api.nvim_get_autocmds`
- `vim.api.nvim_do_autocmd`

closes #17552

gpanders commented on src/nvim/api/autocmd.c:
```diff
@@ -57,15 +57,26 @@ Array nvim_get_autocmds(Dict(get_autocmds) *opts, Error *err)
     Object v = opts->group;
-    if (v.type != kObjectTypeString) {
-      api_set_error(err, kErrorTypeValidation, "group must be a string.");
-      goto cleanup;
-    }
+    switch (v.type) {
+    case kObjectTypeString:
```
`kObjectTypeNil` is just another type, so you can include it in the switch statement:

```c
switch (opts->group.type) {
case kObjectTypeString:
  // handle string case
  break;
case kObjectTypeInteger:
  // handle int case
  break;
case kObjectTypeNil:
  // do nothing
  break;
default:
  // error
  api_set_error(...);
  break;
}
```

gpanders commented on src/nvim/api/autocmd.c:
```diff
@@ -349,22 +360,29 @@ Integer nvim_create_autocmd(uint64_t channel_id, Object event, Dict(create_autoc
-      api_set_error(err,
-                    kErrorTypeException,
-                    "invalid augroup: %s", v->data.string.data);
-
-      goto cleanup;
+    switch (v->type) {
```
Same comment here about `switch`

gpanders commented on src/nvim/api/autocmd.c:
```diff
@@ -595,18 +613,30 @@ void nvim_do_autocmd(Object event, Dict(do_autocmd) *opts, Error *err)
-      api_set_error(err, kErrorTypeValidation, "'group' must be a string");
-      goto cleanup;
-    }
+    Object *v = &opts->group;
```
Same here

muniter reviewed (commented):
I think the function chosen for this [`augroup_name`](https://github.com/neovim/neovim/blob/master/src/nvim/autocmd.c#L462) is not ideal to use this heavily in the API.

The function [iterates over all the map (augroup_map) values](https://github.com/neovim/neovim/blob/master/src/nvim/autocmd.c#L493) to find the given ID. So every call to one of the autocmd API functions which we expect quite a few will do these iterations.

We currently have:

- [`Map(augrup_name, augroup_id)`](https://github.com/neovim/neovim/blob/master/src/nvim/autocmd.c#L110)

For this feature I think it will be better to do refactoring and have an extra `Map(agroup_id, augrup_name)` for fast and simple lookups.

tjdevries commented:
I'm not sure how much of a big deal it is that we have to iterate over the list, but we could maybe add two maps and maintain both of them, that's not too crazy to do.

dhruvmanila commented:
I adopted the logic from `nvim_del_augroup_by_id`:

https://github.com/neovim/neovim/blob/770b9f3c3b2d236ea3e1a480a2b06a9b8846508f/src/nvim/api/autocmd.c#L550-L555

I thought of adding a map as well, but the only thing needed to be done was to check whether the augroup with given id exists or not.

I guess adding a reverse map should not be much work but then I think the following changes make sense:
- Update `augroup_name` to use the map instead of iteration
- Add a function to delete entry from the map similar to `augroup_map_del`
- Maybe change names to reflect the map used?
  - `augroup_map` -> `augroup_name_map` (reverse: `augroup_id_map`)
  - `augroup_map_del` -> `augroup_name_map_del` (reverse: `augroup_id_map_del`)

tjdevries committed:
refactor: add two maps for fast lookups

tjdevries commented:
I just threw together a commit for the double map thing, seems to work (didn't check a lot). It involves doing a few goofy things w/ map internals, so I figured I would just push that and you can play w/ it as necessary. Or you can delete my commit if you hate it :laughing:

tjdevries commented:
I have a suspicion that this might be doing some things as well:

https://github.com/neovim/neovim/issues/17561#issuecomment-1055638003

I can try and track this down in a different PR if you want (and we could drop my commit from this PR as well if we wanted, and I'll do this in a followup PR)

dhruvmanila commented:
I did the same changes locally, except for the variable names and added a new function `augroup_id_map_del` for the second map, which I would prefer instead of one function.

I'll just fix some errors in your changes and we can see where the tests are failing, as they were when I ran it locally.

dhruvmanila commented:
It seems that the old tests are failing after introducing the reverse map: `make oldtest TEST_FILE=test_autocmd.vim`
* `Test_augroup_deleted`
* `Test_augroup_warning`

Both seems to be related to the [`augroup_del` function](https://github.com/neovim/neovim/blob/master/src/nvim/autocmd.c#L407).

dhruvmanila commented:
This is weird: I don't know why the tests are failing because running the test scenarios manually in `nvim --clean` is doing exactly what it's suppose to be doing.

<details><summary><code>Test_augroup_deleted</code> code</summary>

https://github.com/neovim/neovim/blob/0e02a5f10909e3be186c1531dc3ca75dbcbf0e33/src/nvim/testdir/test_autocmd.vim#L381-L390

</details>

Here, line 388 is failing, but running the same commands gives me the expected output:

<img width="511" alt="Screenshot 2022-03-02 at 15 15 50" src="https://user-images.githubusercontent.com/67177269/156336857-8d31664e-4a20-40f9-a332-acc99d86ea78.png">

<details><summary><code>Test_augroup_warning</code> code</summary>

https://github.com/neovim/neovim/blob/0e02a5f10909e3be186c1531dc3ca75dbcbf0e33/src/nvim/testdir/test_autocmd.vim#L335-L362

</details>

Here, line 344 and 349 are failing, but, again, running the same commands gives me the expected output

<img width="526" alt="Screenshot 2022-03-02 at 15 23 41" src="https://user-images.githubusercontent.com/67177269/156338234-574d3ee8-6bf3-4c9a-b9d0-e074828525a8.png">

dhruvmanila committed:
fix: delete augroup info from id->name map

When in "stupid_legacy_mode", the value in name->id map would be updated
to `AUGROUP_DELETED`, but the entry would still remain in id->name. This
would create a problem in `augroup_name` function which would return the
name of the augroup instead of `--DELETED--`.

The id->name map is only used for fast loopup in `augroup_name` function
so there's no point in keeping the entry of deleted augroup in it.

dhruvmanila commented:
Hey @tjdevries, I've fixed the remaining issues, this is good to go. Once it's approved I will squash the commits 👍

tjdevries commented on src/nvim/autocmd.c:
```diff
@@ -105,17 +105,28 @@ static char_u *old_termresponse = NULL;
   api_free_string(key);
 }
 
+static void augroup_id_map_del(int id)
```
I'd rather have only one function and pass an ID less than 0 to it to skip deleting the ID from the map. I don't like having two functions here, since it's very easy to forget to use both in the future if we have to make more modifications.

dhruvmanila replied:
Actually, I want the entry to be deleted from id->name map, and _not_ from name->id map.

I could pass in a `NULL` for `name` parameter and put a conditional for it.

dhruvmanila commented on src/nvim/api/autocmd.c:
```diff
@@ -55,19 +55,27 @@ Array nvim_get_autocmds(Dict(get_autocmds) *opts, Error *err)
+      break;
+    case kObjectTypeString:
+      group = augroup_find(opts->group.data.string.data);
+      if (group < 0) {
```
No: 

https://github.com/neovim/neovim/blob/3800615da9eaf9e8b26d9040c882c74084d688e4/src/nvim/autocmd.c#L950-L951

dhruvmanila commented:
The test failure does not seem to be related to this PR: https://github.com/neovim/neovim/runs/5442913716?check_suite_focus=true#step:13:1197

tjdevries commented:
Thanks, nice work :)
